### PR TITLE
[TEP-0096]: Pipelines V1 Scope

### PR DESCRIPTION
**Pull request by lbernick**
This TEP adds use cases for Tekton and criteria for scoping what work should and should not be prioritized for inclusion in a V1 API. It also adds a stability policy for metrics.

Specific features, and whether they will be in scope for V1, will be proposed in a subsequent commit.

Co-authored-by: Jerop Kipruto jerop@google.com @jerop 
Co-authored-by: Priti Desai pdesai@us.ibm.com @pritidesai 

/kind tep

### CODE REVIEW

**Review by dibyom — APPROVED** [review]
I think this incorporates all the feedback from the API WG today!

/approve

**Review by vdemeester — APPROVED** [review]

File: teps/0096-pipelines-v1-api.md [hunk 1]
```diff
@@ -75,6 +76,11 @@ A deprecated API version must continue to be supported for a length of time base
 | Alpha       | 0 releases                    | 0 months                  |
 | Beta        | 9 months or 3 releases        | 9 months                  |
 
+Kubernetes also defines the following stability policy for metrics:
+- Alpha metrics must function for 0 releases total and 0 releases after their announced deprecation.
+- Stable metrics must function for 4 releases or 12 months, and 3 releases or 9 months after their announced deprecation.
+Tekton does not define stability policies for metrics.
```
Comment (pritidesai) [inline teps/0096-pipelines-v1-api.md:82]: NIT:I don't see it as something we should call out. The proposal includes a statement saying "Tekton follows the Kubernetes stability policy for metrics"  which is sufficient and clear.

File: teps/0096-pipelines-v1-api.md [hunk 2]
```diff
@@ -123,19 +131,59 @@ of TEP-0033: Feature Gates, Tekton's policy allows a behavior change to be updat
...
-### API Definition
-
 ## Use Cases
 
-### V1 Supported Use Cases
+Most Tekton use cases are supported by features from multiple Tekton projects, meaning that there are relatively few use cases for
+Pipelines as an independent project. The following use cases guide Tekton feature work, and a Pipelines V1 API should stabilize
+the features critical to supporting them.
+
+### Use Cases prioritized for Tekton
```
Comment (pritidesai) [inline teps/0096-pipelines-v1-api.md:140]: this is a great start, thank your for adding these!

Comment (pritidesai) [reply teps/0096-pipelines-v1-api.md:140]: this is a limited set of use cases but a great start, do we envision adding more in next iterations? 

I would start this paragraph with:

Tekton is a cloud-native CI/CD platform and is designed to support modern CI/CD use cases, some of them are listed here:

File: teps/0096-pipelines-v1-api.md [hunk 3]
```diff
@@ -123,19 +131,59 @@ of TEP-0033: Feature Gates, Tekton's policy allows a behavior change to be updat
...
+- As an application engineer, I want to run tests when new pull requests are created and automatically
+merge only changes that have passed these tests, so we are always in a releaseable state.
```
Comment (pritidesai) [inline teps/0096-pipelines-v1-api.md:142]: What do you think of rewording this to - As an application developer, I want to design a continuous integration process for my application. The continuous integration includes running tests (unit/integration/etc) when new pull requests are created. And automatically merge only changes that have passed these tests and collected enough reviews such that we are always in a releasable state.

File: teps/0096-pipelines-v1-api.md [hunk 4]
```diff
@@ -123,19 +131,59 @@ of TEP-0033: Feature Gates, Tekton's policy allows a behavior change to be updat
...
+- As an application engineer, I want to be able to build, publish, and deploy my application to staging and prod environments.
+-  As a platform engineer, I want to provide functionality that’s not directly available in the Tekton Pipelines API to application engineers in my organization.
```
Comment (pritidesai) [inline teps/0096-pipelines-v1-api.md:144]: This sounds a little ambiguous. I think this is inferring to custom task. I would rephrase this as - As a CI/CD platform provider, I have an ability to extend the Tekton Pipelines API if my use case is not directly supported/available by the pipelines.

File: teps/0096-pipelines-v1-api.md [hunk 5]
```diff
@@ -123,19 +131,59 @@ of TEP-0033: Feature Gates, Tekton's policy allows a behavior change to be updat
...
+
+### Use Cases not prioritized
+The following use cases may currently be supported by Tekton, but features that solely address these use cases should not be
+V1 API blockers.
+
+- Non-CI/CD workflows, such as machine learning use cases
```
Comment (pritidesai) [inline teps/0096-pipelines-v1-api.md:150]: I don't see this explicit reference to "machine learning" necessary here. This could create conflicts since it's explicitly spelled out. Machine learning is a huge space and Tekton Pipelines is a CI/CD platform. Just having it listed as non-CI/CD workflows is enough.

**Comment by pritidesai** [comment]
thanks a bunch @lbernick for all the details. I have left a few comments otherwise looks great.

/approve

File: teps/0096-pipelines-v1-api.md (hunk 1 above)
Comment (lbernick) [reply teps/0096-pipelines-v1-api.md]: good point, done!

File: teps/0096-pipelines-v1-api.md (hunk 2 above)
Comment (lbernick) [reply teps/0096-pipelines-v1-api.md:139]: Reworded to make it more clear that this isn't intended to be an exhaustive list of use cases.

File: teps/0096-pipelines-v1-api.md (hunk 5 above)
Comment (lbernick) [reply teps/0096-pipelines-v1-api.md]: done

File: teps/0096-pipelines-v1-api.md (hunk 3 above)
Comment (lbernick) [reply teps/0096-pipelines-v1-api.md:142]: I am leaning towards trying to keep use cases concise because they haven't actually helped with scoping as much as I originally expected. Each of these use cases could be expanded to a large number of features, but I think we should keep "new features we want in v1" as short of a list as possible. I'd prefer to keep as is.

**Review by jerop — APPROVED** [review]
thanks @lbernick 🎉
